From 22f15d4137cb5d090f13fc5d9093dc3085dce67b Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Micha=C5=82=20Kosek?= <mihao@users.noreply.github.com>
Date: Tue, 9 Jul 2019 15:34:13 +0200
Subject: [PATCH] Do not "Bind" docker "To" containerd.

relates to https://github.com/docker/for-linux/issues/678

When using the BindTo directive, Docker is permanently stopped by systemd
when containerd is temporarily killed and restarted;

Using `Requires` achieves mostly the same, but defines a weaker dependency;

https://www.freedesktop.org/software/systemd/man/systemd.unit.html#Requires=

> Requires=
>
> .. If this unit gets activated, the units listed will be activated as well.
> If one of the other units fails to activate, and an ordering dependency
> After= on the failing unit is set, this unit will not be started. Besides,
> with or without specifying After=, this unit will be stopped if one of the
> other units is explicitly stopped.

We may want to look into using `Wants=` instead of `Requires=`, because
that allows docker to continue running if containerd is restarted, quoting
the systemd documentation:

> Often, it is a better choice to use Wants= instead of Requires= in order
> to achieve a system that is more robust when dealing with failing services.

Given that docker will likely still fail if the containerd socket is not
present, startup will fail if containerd is not running, but if containerd
is restarted, the docker daemon may be able to try reconnecting.

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>
---
 systemd/docker.service | 3 +--
 1 file changed, 1 insertion(+), 2 deletions(-)

Origin: upstream, https://github.com/docker/docker-ce-packaging/pull/508
Bug: https://github.com/docker/for-linux/issues/678
Ubuntu-Bug: https://bugs.launchpad.net/ubuntu/+source/containerd/+bug/1870514
Reviewed-By: Bryce Harrington <bryce@canonical.com>
Description: This is a backport of upstream's patch, with the path to
 the docker.service modified and using Wants= rather than Requires=, for
 reasons outlined above (Requires= did not work).  The After= target
 also differs from upstream (we don't carry commit 36bb01538.)
Last-Updated: 2020-12-02

diff --git a/systemd/docker.service b/systemd/docker.service
index 9c1d9e6d37..0a6a3064a4 100644
--- a/components/packaging/systemd/docker.service
+++ b/components/packaging/systemd/docker.service
@@ -1,10 +1,10 @@
 [Unit]
 Description=Docker Application Container Engine
 Documentation=https://docs.docker.com
-BindsTo=containerd.service
 After=network-online.target firewalld.service containerd.service
 Wants=network-online.target
 Requires=docker.socket
+Wants=containerd.service
 
 [Service]
 Type=notify
